source/common/shared_pool/shared_pool.h

```diff
@@ -30,60 +31,63 @@ namespace SharedPool {
  * There is also a need to ensure that the thread where ObjectSharedPool's destructor is also in the
  * main thread, or that ObjectSharedPool destruct before the program exit
  */
-template <typename T, typename HashFunc = std::hash<T>,
+template <typename T, typename HashFunc = std::hash<T>, typename EqualFunc = std::equal_to<T>,
           class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
-class ObjectSharedPool : public Singleton::Instance,
-                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
-                         NonCopyable {
+class ObjectSharedPool
+    : public Singleton::Instance,
+      public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc, EqualFunc>>,
+      NonCopyable {
 public:
   ObjectSharedPool(Event::Dispatcher& dispatcher)
       : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
 
-  void deleteObject(const size_t hash_key) {
+  void deleteObject(T* ptr) {
     if (std::this_thread::get_id() == thread_id_) {
-      // There may be new inserts with the same hash value before deleting the old element,
-      // so there is no need to delete it at this time.
-      if (object_pool_.find(hash_key) != object_pool_.end() &&
-          object_pool_[hash_key].use_count() == 0) {
-        object_pool_.erase(hash_key);
+      if (auto iter = object_pool_.find(ptr); iter != object_pool_.end()) {
+        // It is possible that the entry in object_pool_ corresponds to a
+        // different weak_ptr, due to a race condition in a shared_ptr being
```

Comment:
This comment confused me but I think I see the point. It isn't that it's a *different* weak_ptr; it's that we have kind of deferred the destruction of a shared_ptr whose ref-count goes to zero with the callback, and in the meantime another reference to the shared_ptr may be created, making its use-count transition from 0 to 1 while waiting for the post.